tests/TestKdbx4.cpp

```diff
@@ -0,0 +1,189 @@
+    QCOMPARE(entry->attachments()->value("Format400"), QByteArray("Format400\n"));
+}
+
+void TestKdbx4::testFormat400Upgrade()
+{
+    QFETCH(Uuid, kdfUuid);
+    QFETCH(Uuid, cipherUuid);
+    QFETCH(quint32, expectedVersion);
+
+    QScopedPointer<Database> sourceDb(new Database());
+    sourceDb->metadata()->setName("Wubba lubba dub dub");
+    QCOMPARE(sourceDb->kdf()->uuid(), KeePass2::KDF_AES_KDBX3);    // default is legacy AES-KDF
+
+    CompositeKey key;
+    key.addKey(PasswordKey("I am in great pain, please help me!"));
+    sourceDb->setKey(key, true, true);
+
+    QBuffer buffer;
+    buffer.open(QBuffer::ReadWrite);
+
+    // upgrade to KDBX 4 by changing KDF and Cipher
+    sourceDb->changeKdf(KeePass2::uuidToKdf(kdfUuid));
+    sourceDb->setCipher(cipherUuid);
+    KeePass2Writer writer;
+    writer.writeDatabase(&buffer, sourceDb.data());
+    if (writer.hasError()) {
+        QFAIL(qPrintable(QString("Error while writing database: %1").arg(writer.errorString())));
+    }
+
+    // read buffer back
+    buffer.seek(0);
+    KeePass2Reader reader;
+    QScopedPointer<Database> targetDb(reader.readDatabase(&buffer, key));
+    if (reader.hasError()) {
+        QFAIL(qPrintable(QString("Error while reading database: %1").arg(reader.errorString())));
+    }
+
+    QVERIFY(targetDb->rootGroup());
+    QCOMPARE(targetDb->metadata()->name(), sourceDb->metadata()->name());
+
+    QCOMPARE(reader.version(), expectedVersion);
+    QCOMPARE(targetDb->kdf()->uuid(), sourceDb->kdf()->uuid());
+    QCOMPARE(targetDb->cipher(), cipherUuid);
+}
+
+void TestKdbx4::testFormat400Upgrade_data()
+{
+    QTest::addColumn<Uuid>("kdfUuid");
+    QTest::addColumn<Uuid>("cipherUuid");
+    QTest::addColumn<quint32>("expectedVersion");
+
+    auto constexpr kdbx3 = KeePass2::FILE_VERSION_3_1 & KeePass2::FILE_VERSION_CRITICAL_MASK;
+    auto constexpr kdbx4 = KeePass2::FILE_VERSION_4   & KeePass2::FILE_VERSION_CRITICAL_MASK;
+
+    QTest::newRow("Argon2           + AES")      << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_AES      << kdbx4;
+    QTest::newRow("AES-KDF          + AES")      << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_AES      << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + AES")      << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_AES      << kdbx3;
+    QTest::newRow("Argon2           + ChaCha20") << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_CHACHA20 << kdbx4;
+    QTest::newRow("AES-KDF          + ChaCha20") << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_CHACHA20 << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + ChaCha20") << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_CHACHA20 << kdbx3;
+    QTest::newRow("Argon2           + Twofish")  << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_TWOFISH  << kdbx4;
+    QTest::newRow("AES-KDF          + Twofish")  << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_TWOFISH  << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + Twofish")  << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_TWOFISH  << kdbx3;
```

Comment:
@phoerious I wasn't aware of QT's data driven testing capabilities... this is neat!

Reply:
It's pretty amazing to work with.